# **BTC1895 Final Project** 
#### By: Hasan Abdo, Tenzin Gyaltsen, and Neha Kodali
#### April 22, 2025

NOTE: If running cells, please run cells in order due to shared variable names between sections.

### **Import Packages**

In [1]:
# Import required packages for data loading, preprocessing and EDA.
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import plotly.express as px
import numpy as np
import plotly.graph_objects as go
from plotly.subplots import make_subplots

# Import required packages for multiple imputation.
from sklearn.experimental import enable_iterative_imputer 
from sklearn.impute import IterativeImputer
# Other packages for machine learning models will be imported later.

### **Load and View Data**

In [2]:
# Load data set.
df = pd.read_csv('medication_demand_data.csv')

In [3]:
# Inital view of data.
print(df.head())
print('-' * 75)
# Check data types and non-null counts. Nulls exist! 7300 observations in total.
print(df.info())
print('-' * 75)
# Descriptive statistics for numerical variables.
print(df.describe())
print('-' * 75)
# Count number of nulls in each variable.
print(df.isnull().sum())
print('-' * 75)
# Checking if any other commonly interpreted null values.
print((df == '').sum())
print('-' * 75)
print((df == ' ').sum())
print('-' * 75)
print((df == 'NA').sum())
# Only nulls in this dataset.


         Date      Medication   Sales     Region  Temperature  Humidity  \
0  2023-01-01     Cold Relief  2084.0    Toronto    -3.500000      59.3   
1  2023-01-01     Cold Relief  1372.0  Vancouver     6.600000      57.7   
2  2023-01-01     Cold Relief  1611.0   Montreal   -26.165964      52.9   
3  2023-01-01     Cold Relief  1671.0    Calgary   -13.900000      57.0   
4  2023-01-01  Allergy Relief   351.0    Toronto    -4.400000      50.2   

   Flu_Cases  Pollen_Count  Google_Trends  Marketing_Spend         Holiday  
0      312.0           6.0           95.8          3230.23  New Year's Day  
1      290.0           9.0           75.3          2696.15  New Year's Day  
2      298.0           7.0           74.6          3033.28  New Year's Day  
3      309.0           6.0           77.1          3246.76  New Year's Day  
4      302.0           5.0           48.8           909.67  New Year's Day  
---------------------------------------------------------------------------
<class 'pan

### **Exploratory Data Analysis (High-Level)**

In [4]:
# Generate a quick and visually compact exploratory data analysis (EDA), first at a high-level just showing monovariate distributions.
# Plots are made via 'plotly' for both numeric and categorical variables.
# Histograms and bar charts are tiled into subplots to make them dashboard-ready. 
# Easy to scan and space-efficient, especially for larger datasets.

# Define explicit lists of numeric and categorical variables to include in EDA.
numeric_vars = ['Sales', 'Temperature', 'Humidity', 'Flu_Cases',
                'Pollen_Count', 'Google_Trends', 'Marketing_Spend']
categorical_vars = ['Medication', 'Region', 'Holiday']

# Define function to create tiled histograms for numeric variables.
def eda_numeric(df):
    # Set number of columns (tiles) per row.
    cols = 2
    # Calculate number of rows using ceiling division.
    rows = -(-len(numeric_vars) // cols)
    # Create a subplot layout for all numeric variables.
    fig = make_subplots(rows=rows, cols=cols, subplot_titles=numeric_vars)
    # Initialize counter to track subplot placement.
    idx = 0
    # Loop through numeric variables.
    for i, col in enumerate(numeric_vars):
        # Skip if column doesn't exist in the dataframe.
        if col not in df.columns:
            continue
        # Update index and determine subplot position.
        idx += 1
        row = (idx - 1) // cols + 1
        col_pos = (idx - 1) % cols + 1
        # Create histogram and add to subplot.
        hist = go.Histogram(x=df[col], name=col)
        fig.add_trace(hist, row=row, col=col_pos)
    # Configure overall layout and axes.
    fig.update_layout(
        title_text='Histograms of Numeric Variables',
        height=300 * rows,
        showlegend=False
    )
    fig.update_xaxes(title_text='Value')
    fig.update_yaxes(title_text='Count')
    # Display figure.
    fig.show()


# Define function to create tiled bar charts for categorical variables.
def eda_categorical(df):
    # Set number of columns (tiles) per row.
    cols = 2
    # Calculate number of rows needed.
    rows = -(-len(categorical_vars) // cols)
    # Create a subplot layout for all categorical variables.
    fig = make_subplots(rows=rows, cols=cols, subplot_titles=categorical_vars)
    # Initialize a counter to track subplot position.
    idx = 0
    # Loop through categorical variables.
    for i, col in enumerate(categorical_vars):
        # Skip if column doesn't exist
        if col not in df.columns:
            continue
        # Update the index and determine subplot position.
        idx += 1
        row = (idx - 1) // cols + 1
        col_pos = (idx - 1) % cols + 1
        # Replace missing values and convert to string.
        filled = df[col].fillna('Missing').astype(str)
        # Calculate value counts.
        counts = filled.value_counts().reset_index()
        counts.columns = ['Category', 'Count']
        # Create bar chart and add to subplot.
        bar = go.Bar(x=counts['Category'], y=counts['Count'], name=col)
        fig.add_trace(bar, row=row, col=col_pos)
        # Rotate x-axis labels for readability.
        fig.update_xaxes(tickangle=45, row=row, col=col_pos)
    # Configure overall layout and axes.
    fig.update_layout(
        title_text='Bar Charts of Categorical Variables',
        height=300 * rows,
        showlegend=False
    )
    fig.update_yaxes(title_text='Count')
    # Display figure.
    fig.show()

# Run both numeric and categorical EDA visualizations.
eda_numeric(df)
eda_categorical(df)

### NOTE: 
The purpose of these high-level "distribution" or "count"-based EDA plots is to provide a quick visual understanding of the variables' structure and spread. These plots are not intended to generate business insights at this stage, but rather to inform and guide future preprocessing and analytical decisions.

### **Exploratory Data Analysis (For Business Insights)**

In [5]:
# Make a separate copy of the original dataset to avoid modifying it during EDA. Some steps require creating new variables.
df_eda = df.copy()
# Round Google Trends scores to the nearest whole number.
# This reduces noise and helps group similar search volumes together.
df_eda['GT_rounded'] = df_eda['Google_Trends'].round()

# Group by rounded Google Trends score and Medication type.
# Then calculate the average Sales for each combination.
# This simplifies the data and makes it easier to visualize general patterns.
agg_line = df_eda.groupby(['GT_rounded', 'Medication'], observed=True).agg(
    Avg_Sales=('Sales', 'mean')
).reset_index()
# Create a scatter plot showing the average Sales vs rounded Google Trends.
# Each point represents the mean sales for a specific Google Trends score and medication.
# A fitted linear regression line (trendline='ols') is added per medication to show the trend and rate of change.
fig = px.scatter(
    agg_line,
    x='GT_rounded',
    y='Avg_Sales',
    color='Medication',
    trendline='ols',
    title='Average Sales vs Google Trends (with Fitted Trendlines)',
    labels={'GT_rounded': 'Google Trends', 'Avg_Sales': 'Average Sales'}
)

# Center the title and display the plot.
fig.update_layout(title_x=0.5)
fig.show()

This plot supports the idea that Google Trends can serve as a leading indicator of medication demand, as all fitted trendlines show a positive slope. This suggests that higher search interest is generally associated with increased sales. 

Medications tied to seasonal or outbreak-driven symptoms, such as cough suppressants, cold relief, and allergy medications which exhibit the steepest slopes, indicating that their sales may be more sensitive to changes in public interest.

A higher slope reflects a stronger rate of increase in sales with rising search volume, making these categories potentially more predictable or responsive in demand forecasting efforts.

In [6]:
# Instead of plotting raw Marketing Spend (which is too noisy), we bin it into 10 quantiles.
# This groups spend values into 10 buckets based on distribution, not fixed ranges, but balanced by count.
# We use qcut (quantile cut) to do that, and drop duplicates just in case some bins overlap due to identical values.
df_eda['Spend_bin'] = pd.qcut(df_eda['Marketing_Spend'], q=10, duplicates='drop')

# Now we aggregate by bin and Medication — we are getting the average Sales and average Spend per combination.
# This simplifies the data and gives us a much cleaner way to look at how sales behave across increasing spend levels.
agg_bin = df_eda.groupby(['Spend_bin', 'Medication'], observed=True).agg(
    Avg_Sales=('Sales', 'mean'),
    Avg_Spend=('Marketing_Spend', 'mean')
).reset_index()

# We convert the bin intervals (which are unappealing like "[1354.0, 1785.5]") into strings for tooltips only.
# The x-axis will use Avg_Spend instead, so the spacing actually reflects the amount of money spent.
agg_bin['Spend_bin'] = agg_bin['Spend_bin'].astype(str)

# Now we plot where x-axis is Avg_Spend, y-axis is Avg_Sales, and the color shows the Medication type.
# This gives us a clean line chart with markers, showing how sales change with increasing spend.
# Much easier to interpret than a messy scatter (which is what we tried to improve from our old graphs).
fig = px.line(
    agg_bin,
    x='Avg_Spend',
    y='Avg_Sales',
    color='Medication',
    markers=True,
    title='Average Sales by Binned Marketing Spend and Medication',
    labels={'Avg_Spend': 'Marketing Spend (Avg of Bin)', 'Avg_Sales': 'Average Sales'}
)

# Center the title and show the figure.
fig.update_layout(title_x=0.5)
fig.show()

This plot illustrates how average sales change across increasing levels of marketing spend, grouped into ten bins to smooth out noise and highlight trends.

Cold Relief and Allergy Relief stand out as the most responsive to marketing investment. Their average sales rise significantly with higher spending, particularly in the upper bins. Cold Relief shows a strong lift early on but begins to level off, while Allergy Relief continues to climb more steadily across the full range.

In contrast, Pain Relief and Fever Reducer show more moderate responsiveness. Their sales increase slightly with higher spend but tend to plateau, indicating a point of diminishing returns.

Cough Suppressant displays a relatively flat trend throughout and even appears to decline at higher spending levels. This suggests that beyond a certain threshold, additional marketing may not translate into increased sales for that category, possibly due to saturation or consumer limits.

Overall, this chart reveals that not all medications respond equally to marketing investment. Strategic spending on products like Cold Relief and Allergy Relief could offer better returns, while others may require a more cautious approach to avoid overspending where it has limited effect.

In [7]:
# Make sure Date is in datetime format and extract Month + Season.
df['Date'] = pd.to_datetime(df['Date'])
df['Month'] = df['Date'].dt.month_name()
# Map each month to its season.
season_map = {
    'January': 'Winter', 'February': 'Winter', 'March': 'Spring', 'April': 'Spring',
    'May': 'Spring', 'June': 'Summer', 'July': 'Summer', 'August': 'Summer',
    'September': 'Fall', 'October': 'Fall', 'November': 'Fall', 'December': 'Winter'
}
df['Season'] = df['Month'].map(season_map)

# Group by Medication and Season, calculate average sales.
med_month = df.groupby(['Medication', 'Season'], observed=True)['Sales'].mean().reset_index()
# Pivot into a matrix format for heatmap.
heatmap_data = med_month.pivot(index='Medication', columns='Season', values='Sales')
# Reorder seasons manually for better visual logic.
season_order = ['Winter', 'Spring', 'Summer', 'Fall']
heatmap_data = heatmap_data[season_order]

# Heatmap using plotly.
fig = px.imshow(
    heatmap_data,
    text_auto='.1f',
    color_continuous_scale='YlOrRd',
    labels=dict(x='Season', y='Medication', color='Avg Sales'),
    title='Medication Sales by Season (Heatmap)'
)
# Increase figure size for better visibility.
fig.update_layout(
    title_x=0.5,
    width=900,
    height=500
)

# Display heatmap.
fig.show()

This heatmap shows the average sales of each medication across the four seasons, revealing distinct seasonal patterns in demand.

Allergy Relief peaks sharply in the Spring, which aligns with the start of allergy season. Its sales are significantly lower in all other seasons, especially Winter and Fall.

Cold Relief and Cough Suppressant show the opposite trend, with their highest sales occurring in Winter, tapering off as the year progresses. This is consistent with cold and flu season occurring during colder months.

Fever Reducer sees relatively high sales in Winter but maintains moderate demand throughout the year, suggesting more consistent usage or broader symptom application.

Pain Relief shows the most stable sales across seasons, with very little fluctuation. This likely indicates year-round demand unrelated to seasonality.
This seasonal breakdown helps identify when different medications are most in demand. It suggests that marketing and supply strategies (for businesses) should be seasonally adjusted.

In [8]:
# Define correct chronological month order.
month_order = ['January', 'February', 'March', 'April', 'May', 'June',
               'July', 'August', 'September', 'October', 'November', 'December']
# Group by Month and Region, then calculate average sales.
region_over_time = df.groupby(['Month', 'Region'], observed=True)['Sales'].mean().reset_index()
# Make sure months are in the right order.
region_over_time['Month'] = pd.Categorical(region_over_time['Month'], categories=month_order, ordered=True)
region_over_time = region_over_time.sort_values('Month')

# Create a line plot of average sales by region over time.
fig = px.line(
    region_over_time,
    x='Month',
    y='Sales',
    color='Region',
    markers=True,
    title='Sales Trends by Region (Line Plot)',
    labels={'Sales': 'Average Sales'}
)
# Setting layout options.
fig.update_layout(
    title_x=0.5,
    xaxis_title='Month',
    yaxis_title='Average Sales',
    legend_title='Region',
)

# Display the plot.
fig.show()

This plot shows that average sales peak during winter months (January, February, and December) across all regions, likely due to seasonal factors like cold and flu activity.

Sales begin to decline through March to May, but the lowest levels are reached in the summer, especially in June, July, and August.

After that, sales gradually rise again through fall, with a sharp increase in December.

Toronto shows the highest average sales overall, while Vancouver tends to trend slightly lower. Despite differences in magnitude, all regions seem to follow a similar seasonal sales pattern.

In [9]:
# Only plot Allergy Relief here because it is the only medication directly affected by pollen levels.
# Including other medications would dilute the relationship between pollen levels and sales, and may make the scatter plot confusing.
allergy_df = df_eda[df_eda['Medication'] == 'Allergy Relief']
# Create a scatter plot with a fitted linear trendline (OLS) or Ordinary Least Squares.
fig = px.scatter(
    allergy_df,
    x='Pollen_Count',
    y='Sales',
    trendline='ols',
    title='Allergy Relief Sales vs Pollen Count',
    labels={'Pollen_Count': 'Pollen Count', 'Sales': 'Sales'},
    opacity=0.6
)

# Center title and show plot.
fig.update_layout(title_x=0.5)
fig.show()

This scatter plot shows a positive relationship between pollen count and Allergy Relief sales. As pollen levels rise, sales generally increase as well. The upward trendline indicates that higher pollen exposure could be associated with greater demand for Allergy Relief products. Despite some variation at extreme values, the overall pattern suggests a seasonal effect where allergy-related sales respond to environmental triggers (pollen).

### **Data Pre-processing**

In [10]:
# Preprocessing steps. Create a fresh copy of the data.
df_preprocessed = df.copy()
# Binarize 'Holiday' column. First fill null values with an empty string for later recognition.
df_preprocessed['Holiday'] = df_preprocessed['Holiday'].fillna('')
# Define a function to classify the holiday as either a holiday or not.
def binarize_holiday(x):
    if x.strip() != '':
        return 'Holiday'
    else:
        return 'Not a Holiday'
# Apply the function to each value in the 'Holiday' column.
df_preprocessed['Holiday'] = df_preprocessed['Holiday'].apply(binarize_holiday)

# Round down all values in the 'Sales' column as they intended to be whole number integers but some are floats. 
df_preprocessed['Sales'] = np.floor(df_preprocessed['Sales']).astype(int)
# View changes for the first few rows to ensure changes have taken effect. 
print(df_preprocessed.head())
# Check that 'Sales' is now integer type instead of float.
print(df_preprocessed.dtypes)

        Date      Medication  Sales     Region  Temperature  Humidity  \
0 2023-01-01     Cold Relief   2084    Toronto    -3.500000      59.3   
1 2023-01-01     Cold Relief   1372  Vancouver     6.600000      57.7   
2 2023-01-01     Cold Relief   1611   Montreal   -26.165964      52.9   
3 2023-01-01     Cold Relief   1671    Calgary   -13.900000      57.0   
4 2023-01-01  Allergy Relief    351    Toronto    -4.400000      50.2   

   Flu_Cases  Pollen_Count  Google_Trends  Marketing_Spend  Holiday    Month  \
0      312.0           6.0           95.8          3230.23  Holiday  January   
1      290.0           9.0           75.3          2696.15  Holiday  January   
2      298.0           7.0           74.6          3033.28  Holiday  January   
3      309.0           6.0           77.1          3246.76  Holiday  January   
4      302.0           5.0           48.8           909.67  Holiday  January   

   Season  
0  Winter  
1  Winter  
2  Winter  
3  Winter  
4  Winter  
Date    

### **Imputation and Checking Overfitting Rule of Thumb**

In [11]:
# Imputation time. Check again number of nulls in each variable.
# Count missing values per column.
missing_counts = df_preprocessed.isnull().sum()
# Calculate percentage of missing values.
missing_percent = (missing_counts / len(df_preprocessed)) * 100
# Combine both into one DataFrame.
missing_table = pd.DataFrame({
    'Missing Values': missing_counts,
    'Percentage (%)': missing_percent.round(2)
})
# Filter only columns with missing data.
missing_table = missing_table[missing_table['Missing Values'] > 0]
# Display sorted by missing count.
print("Missing Values Summary:")
print(missing_table.sort_values(by='Missing Values', ascending=False)) 
print('-' * 50)
# Maximum number of nulls is 381, which is less than 30% of total observations. Proceed with imputation.

# Create imputer object.
imputer = IterativeImputer(random_state=0)
# Fit and impute (multiple imputation) columns with missingness.
df_clean = df_preprocessed.copy()
df_clean[['Temperature', 'Humidity', 'Flu_Cases', 'Pollen_Count', 'Google_Trends', 'Marketing_Spend']] = imputer.fit_transform(
    df_clean[['Temperature', 'Humidity', 'Flu_Cases', 'Pollen_Count', 'Google_Trends', 'Marketing_Spend']])

# Check descriptive statistics again, pre- and post-imputation.
print(df_preprocessed.describe())
print('-' * 50)
print(df_clean.describe())
print('-' * 50)
# Calculate and print a table of the difference in descriptive statistics as a result of multiple imputation.
desc_before = df_preprocessed.describe()
desc_after = df_clean.describe()
desc_diff = desc_after - desc_before
# Print only imputed columns.
columns_of_interest = ['Temperature', 'Humidity', 'Flu_Cases', 'Pollen_Count', 'Google_Trends', 'Marketing_Spend']
print(desc_diff[columns_of_interest])

Missing Values Summary:
                 Missing Values  Percentage (%)
Temperature                 381            5.22
Marketing_Spend             380            5.21
Humidity                    374            5.12
Pollen_Count                370            5.07
Google_Trends               368            5.04
Flu_Cases                   348            4.77
--------------------------------------------------
                                Date        Sales  Temperature     Humidity  \
count                           7300  7300.000000  6919.000000  6926.000000   
mean   2023-07-02 00:00:00.000000256   733.711781     9.500998    67.692249   
min              2023-01-01 00:00:00   197.000000   -28.765964    44.100000   
25%              2023-04-02 00:00:00   438.000000     4.400000    62.100000   
50%              2023-07-02 00:00:00   620.000000    11.000000    67.600000   
75%              2023-10-01 00:00:00   818.000000    17.900000    73.100000   
max              2023-12-31 00:00:00

In [12]:
# Extra check for overfitting rule of thumb. Ideal to have (number of samples in smallest class / number of predictors) be greater or equal to 
# 10-15. We have 13 features (but really 10 since we will only use one of 'Date', 'Month' and 'Season', due to potential multicollinearity, and 
# 'Sales' is the outcome variable).
print(min(
    df_clean['Medication'].value_counts().min(),
    df_clean['Region'].value_counts().min(),
    df_clean['Holiday'].value_counts().min(),
    df_clean['Month'].value_counts().min(),
    df_clean['Season'].value_counts().min()
    )
)
# The smallest class has 200 values (i.e. 'Holiday' in the 'Holiday' column).
print(200/10)
# Equal to 20, which is greater than 15. Passes the rule of thumb!



200
20.0


In [61]:
# Save clean data as a separate CSV file for the dash app.
df_clean.to_csv("clean_medication_data.csv", index=False)

## **Machine Learning Models** 

### **Import Machine Learning Packages**

In [14]:
# Time for some machine learning to predict medication demand. 
# First import necessary packages for all models.

from sklearn.model_selection import train_test_split
from sklearn import metrics
from sklearn.metrics import mean_squared_error, r2_score, mean_absolute_error, root_mean_squared_error
import tensorflow as tf
from tensorflow.keras.metrics import RootMeanSquaredError
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import KFold

from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestRegressor, AdaBoostRegressor
from sklearn.neighbors import KNeighborsRegressor
from sklearn.tree import DecisionTreeRegressor
from xgboost import XGBRegressor
from sklearn.model_selection import GridSearchCV

from tensorflow.keras.metrics import RootMeanSquaredError
from tensorflow import keras
from tensorflow.keras import Input

from statsmodels.tsa.statespace.sarimax import SARIMAX
from datetime import timedelta



#### **Linear Regression**

In [42]:
# 1. Select the categorical columns that we’ll need to encode.
# These are non-numeric features that the model wont be able to understand as-is.
cat_col = ['Medication', 'Region', 'Season', 'Holiday']

# 2. Apply one-hot encoding to convert the categorical variables into numerical format.
# This just creates new columns (0s and 1s) to represent each category — super important for making these usable in the model.
onehotencoder = OneHotEncoder(sparse_output=False)
one_hot_encoded = onehotencoder.fit_transform(df_clean[cat_col])
one_hot_df = pd.DataFrame(one_hot_encoded, columns=onehotencoder.get_feature_names_out(cat_col))

# 3. Dropping the original categorical columns (since we’ve already encoded them), plus Date and Month.
# These aren't useful for our model in their current form.
drop_cols = cat_col + ['Date']
if 'Month' in df_clean.columns:
    drop_cols.append('Month')  
# Put it all together. Keep the rest of the numeric columns and add in the new one-hot encoded ones.
df_sklearn_encoded = pd.concat([
    df_clean.drop(columns=drop_cols).reset_index(drop=True),
    one_hot_df.reset_index(drop=True)
], axis=1)
# Check how the new encoded dataframe looks.
df_sklearn_encoded.head()

Sales  Temperature  Humidity  Flu_Cases  Pollen_Count  Google_Trends  \
0   2084    -3.500000      59.3      312.0           6.0           95.8   
1   1372     6.600000      57.7      290.0           9.0           75.3   
2   1611   -26.165964      52.9      298.0           7.0           74.6   
3   1671   -13.900000      57.0      309.0           6.0           77.1   
4    351    -4.400000      50.2      302.0           5.0           48.8   

   Marketing_Spend  Medication_Allergy Relief  Medication_Cold Relief  \
0          3230.23                        0.0                     1.0   
1          2696.15                        0.0                     1.0   
2          3033.28                        0.0                     1.0   
3          3246.76                        0.0                     1.0   
4           909.67                        1.0                     0.0   

   Medication_Cough Suppressant  ...  Region_Calgary  Region_Montreal  \
0                           0.0  ...             0.0              0.0   
1                           0.0  ...             0.0              0.0   
2                           0.0  ...             0.0              1.0   
3                           0.0  ...             1.0              0.0   
4                           0.0  ...             0.0              0.0   

   Region_Toronto  Region_Vancouver  Season_Fall  Season_Spring  \
0             1.0               0.0          0.0            0.0   
1             0.0               1.0          0.0            0.0   
2             0.0               0.0          0.0            0.0   
3             0.0               0.0          0.0            0.0   
4             1.0               0.0          0.0            0.0   

   Season_Summer  Season_Winter  Holiday_Holiday  Holiday_Not a Holiday  
0            0.0            1.0              1.0                    0.0  
1            0.0            1.0              1.0                    0.0  
2            0.0            1.0              1.0                    0.0  
3            0.0            1.0              1.0                    0.0  
4            0.0            1.0              1.0                    0.0  

[5 rows x 22 columns]

In [43]:
# 4. Define features and target.
# 'Sales' is what we're trying to predict, so we set it as y.
# All other columns will be our input features (X).
X = df_sklearn_encoded.drop(columns='Sales', axis=1)
y = df_sklearn_encoded['Sales']

# 5. Split the data into training and testing sets (80% train, 20% test).
# We will train the model on the training set and evaluate it on the test set.
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

# 6. Scale features using StandardScaler
# Important for models like linear regression that are sensitive to feature scale.
scaler = StandardScaler()
# Fit and transform the training data.
X_train_scaled = scaler.fit_transform(X_train)  
# Use the same transformation on the test data.
X_test_scaled = scaler.transform(X_test)        

# 7. Fit the linear regression model to the scaled training data.
lr = LinearRegression()
lr.fit(X_train_scaled, y_train)
# 8. Print model outputs.
# Coefficient: how much each feature influences the prediction.
# Intercept: the baseline prediction when all features are zero.
# R^2 score: how well the model explains the variation in sales.
print(f"Coefficient (slope): {lr.coef_}")
print(f"Intercept: {lr.intercept_}")
print(f"R-squared Score: {lr.score(X_test_scaled, y_test):.2f}")

Coefficient (slope): [-4.51470879e+01  9.72541322e+00 -2.50983254e-01  1.61908452e+00
  1.98630070e+02  1.68200222e+02 -3.65716988e+13 -3.63037564e+13
 -3.62329496e+13 -3.64675056e+13 -3.64092583e+13 -1.18728868e+13
 -1.18269195e+13 -1.18782479e+13 -1.18159997e+13  1.23795087e+15
  1.24384307e+15  1.23653122e+15  1.23567631e+15 -1.64781805e+15
 -1.64781805e+15]
Intercept: 731.5960298610663
R-squared Score: 0.54


In [44]:
# Use the trained model to make predictions on the test set.
y_pred = lr.predict(X_test_scaled)
# 9. Create a comparison data frame to visually compare predicted vs actual sales.
# This shows the first 6 predictions alongside the true values for a quick check.
comparison_df = pd.DataFrame({
    'Predicted': y_pred[:6],
    'Actual': y_test.values[:6]
})
# Print the comparison table.
print(comparison_df)

    Predicted  Actual
0   486.68978     432
1   379.06478     394
2  1002.68978     710
3  1392.68978    1692
4   910.06478     884
5   685.18978     723


In [45]:
# 10. Evaluate model performance using common regression metrics.
# Mean Squared Error (MSE).
mse = mean_squared_error(y_test, y_pred)
# R-squared Score: measures how well the model explains the variance in the data (1.0 = perfect).
r2 = r2_score(y_test, y_pred)
# Root Mean Squared Error (RMSE): square root of MSE, more interpretable since it's in the same unit as the target (sales).
rmse = np.sqrt(mse) 
# Print out the results.
print(f"Mean Squared Error = {mse:.2f}")
print(f"Root Mean Squared Error = {rmse:.2f}")
print(f"R-squared Score = {r2:.2f}")

# 11. Visualize the model's predictions vs actual sales. 
# Create scatter plot of predicted vs actual values.
fig = go.Figure()
# Add actual vs predicted points.
fig.add_trace(go.Scatter(
    x=y_test,
    y=y_pred,
    mode='markers',
    name='Linear Regression Model',
    opacity=0.5,
    marker=dict(color='blue')
))
# Add reference diagonal line (perfect prediction).
min_val = min(y_test.min(), y_pred.min())
max_val = max(y_test.max(), y_pred.max())
fig.add_trace(go.Scatter(
    x=[min_val, max_val],
    y=[min_val, max_val],
    mode='lines',
    name='Perfect Prediction',
    line=dict(color='red', dash='dash')
))
# Set title and axis labels.
fig.update_layout(
    title='Actual vs Predicted Sales',
    xaxis_title='Actual Sales',
    yaxis_title='Predicted Sales',
    legend_title='',
    width=700,
    height=500,
    template='simple_white'
)

# Show the plot.
fig.show()
# Model doesn't appear to be too accurate, lets try other models. 

Mean Squared Error = 96477.35
Root Mean Squared Error = 310.61
R-squared Score = 0.54


In [46]:
# Extra plot to evaluate performance/fit: Residual plot.
# Residuals help us understand how far off each prediction was from the actual value.
# Calculate residuals.
residuals = y_test - y_pred
# Create the residual plot.
fig = go.Figure()
# Scatter plot of predicted vs residuals.
fig.add_trace(go.Scatter(
    x=y_pred,
    y=residuals,
    mode='markers',
    name='Residuals',
    opacity=0.5,
    marker=dict(color='blue')
))

# Add horizontal reference line at y = 0.
fig.add_trace(go.Scatter(
    x=[y_pred.min(), y_pred.max()],
    y=[0, 0],
    mode='lines',
    name='Zero Error',
    line=dict(color='red', dash='dash')
))
# Layout customization.
fig.update_layout(
    title='Residual Plot (Linear Regression)',
    xaxis_title='Predicted Sales',
    yaxis_title='Residuals (Actual - Predicted)',
    width=700,
    height=500,
    template='simple_white'
)

# Show plot.
fig.show()

For our analysis, we started with linear regression as a baseline to see how well a simple combination of features could predict sales.

After applying one-hot encoding to the categorical variables, we trained the model on 80 percent of the data and tested it on the remaining 20 percent.

The model explained 54 percent of the variance in sales, which makes it a decent starting point. However, the root mean squared error was approximately 311, meaning the model’s predictions were off by about 311 units on average.

From the scatter plot, we can also see that the model struggles more with higher sales values, which suggests it is not capturing non-linear patterns in the data very effectively.

### **Other Machine Learning Models**

In [47]:
# Now that we saw that the linear model doesn't seem to be performing great in terms of its model evaluation, we wanted to try and see if other
# machine learning models can improve performance.
# 1. Define all models.
# Linear Regression.
lr = LinearRegression()
# Random Forest Regressor.
rfr = RandomForestRegressor(n_estimators=70, random_state=0)
# AdaBoost Regressor.
ab = AdaBoostRegressor(n_estimators=70, random_state=0)
# K-Nearest Neighbors Regressor.
knn = KNeighborsRegressor()
# Decision Tree Regressor.
dt = DecisionTreeRegressor(max_depth=7, random_state=0)
# XGBoost Regressor.
xg = XGBRegressor(n_estimators=70, random_state=0)
# Group all models into a dictionary so we can loop through them easily.
clfs = {
    "Linear": lr,
    "RandomForest": rfr,
    "AdaBoost": ab,
    "KNN": knn,
    "DT": dt,
    "XGB": xg  
}

# 2. Define a function to train and evaluate models.
# This will return the R squared score, RMSE, and now also MSE for each model.
def root_mean_squared_error(y_true, y_pred):
    return np.sqrt(mean_squared_error(y_true, y_pred))
def train_and_evaluate(clf, X_train, y_train, X_test, y_test):
    clf.fit(X_train, y_train)
    y_pred = clf.predict(X_test)
    r2 = r2_score(y_test, y_pred)
    rmse = root_mean_squared_error(y_test, y_pred)
    mse = mean_squared_error(y_test, y_pred)
    return r2, rmse, mse

# 3. Loop through models and store results.
results = []
for name, clf in clfs.items():
    r2, rmse, mse = train_and_evaluate(clf, X_train, y_train, X_test, y_test)
    results.append({
        "Model": name,
        "R² Score": round(r2, 3),
        "RMSE": round(rmse, 2),
        "MSE": round(mse, 2)
    })
# Organize the results into a table.
modelmetric_results_df = pd.DataFrame(results).sort_values(by="R² Score", ascending=False)
# Print table.
print(modelmetric_results_df)

# Through this, we find that the Random Forest and XGB models are the best performing models, and could potentially help better predict 
# (more accurately) medication demand (sales). 

          Model  R² Score    RMSE        MSE
5           XGB     0.777  215.56   46464.21
1  RandomForest     0.772  217.95   47501.41
4            DT     0.740  232.93   54257.89
3           KNN     0.554  304.78   92890.29
0        Linear     0.536  310.78   96581.46
2      AdaBoost    -0.052  468.13  219146.17


Instead of randomly making different models, we decided to evaluate models, and based on their performances, chose which one to end up using (for simplicity). We didn't want to make useless models for the sake of it essentially, and wanted to keep our scope more focused. Here we see that both XGB and Random Forest models have the best (similar) performances. 

### **Random Forest**

In [48]:
# We want to take a closer look at random forest models.
# Let's find the best parameters to tune the model.
# 1. Define the initial Random Forest model with a fixed random state.
rfr = RandomForestRegressor(random_state=0)
# 2. Set up a grid of parameters to test during tuning.
# We'll try different numbers of trees and different maximum depths.
param_grid = {
    'n_estimators': [50, 75, 100, 150, 200],
    'max_depth': [5, 10, 15, 20, 25, None]
}
# GridSearchCV is used here to tune the Random Forest model.
# It tests multiple combinations of hyperparameters (like number of trees and tree depth).
# This will help find the best-performing setup based on cross-validated Mean Squared Error (MSE).

# 3. Set up GridSearchCV to perform a cross-validated search over the parameter grid.
grid_search = GridSearchCV(
    estimator=rfr,
    param_grid=param_grid,
    scoring='neg_mean_squared_error',
    cv=10,
    verbose=1,
    n_jobs=-1
)

# 4. Fit the grid search to the training data to find the best parameter combination.
# Note: Random Forest does not require feature scaling since it’s a tree-based model.
# That is why we use X_train not X_train_scaled, and same with Y values.
# As explained here: https://www.kdnuggets.com/2022/07/random-forest-algorithm-need-normalization.html
# And here: https://forecastegy.com/posts/does-random-forest-need-feature-scaling-or-normalization/
# It splits data based on feature thresholds, not distances, so feature magnitude has no effect.
grid_search.fit(X_train, y_train)

# 5. Print the best parameters found by the grid search.
print("Best Parameters:", grid_search.best_params_)
# Print the best MSE (negated to convert back from negative MSE used by GridSearchCV).
print(f"Best MSE: {-grid_search.best_score_:.2f}")

Fitting 10 folds for each of 30 candidates, totalling 300 fits
Best Parameters: {'max_depth': 10, 'n_estimators': 150}
Best MSE: 41870.73


In [49]:
# Refit the best Random Forest model using the full training set.
# This uses the best hyperparameters found by GridSearchCV.
best_rfr = grid_search.best_estimator_
best_rfr.fit(X_train, y_train)

# Make predictions on the test set using the refitted model.
y_pred = best_rfr.predict(X_test)

# Evaluate model performance on the test set.
# MSE: average squared error.
# RMSE: interpretable error in the same units as sales.
# R²: how much variance in sales the model explains.
mse = mean_squared_error(y_test, y_pred)
rmse = root_mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

# Print evaluation metrics.
print(f"Test MSE: {mse:.2f}")
print(f"Test RMSE: {rmse:.2f}")
print(f"Test R² Score: {r2:.3f}")

Test MSE: 47596.78
Test RMSE: 218.17
Test R² Score: 0.771


In [53]:
# Get feature importances from the trained Random Forest model.
# We’re using this to understand which features contributed most to reducing prediction error (MSE).
# These importance scores reflect how much each feature reduced error across all trees in the forest.
# Features that consistently split the data in useful ways during training receive higher importance scores.

# Get feature importances from the trained Random Forest model.
importances = pd.Series(best_rfr.feature_importances_, index=X_train.columns)

# Plot the top 10 most important features.
# This helps us visually identify which inputs the model relied on the most.
top10 = importances.nlargest(10).sort_values()

# Create the horizontal bar chart using plotly.
fig = px.bar(
    x=top10.values,
    y=top10.index,
    orientation='h',
    labels={'x': 'Importance', 'y': 'Feature'},
    title='Top 10 Feature Importances (Random Forest)'
)

# Centre the title.
fig.update_layout(title_x=0.5)

# Show the plot.
fig.show()

# Save the plot for the dash app.
fig.write_html("assets/rf_feature_importance.html")

Looking at the feature importances from the Random Forest model, we found that Marketing Spend had the strongest influence on predictions, contributing around 45% to the model’s decision-making. This highlights a potential strong relationship between marketing efforts and medication sales. Google Trends ranked second, which suggests that online search behavior could be a useful early indicator of consumer demand. 

Interestingly, more traditional seasonal or clinical drivers like flu cases, temperature, and humidity ranked lower, indicating that consumer behaviour and marketing efforts may be more powerful predictors in this context than environmental or epidemiological factors.

In [52]:
# Create a scatter plot to compare actual vs. predicted sales.
# This helps us visually assess how well the model predictions align with the true values.
fig = go.Figure()

# Adding scatter points for predicted vs actual values.
fig.add_trace(go.Scatter(
    x=y_test,
    y=y_pred,
    mode='markers',
    name='Random Forest Predictions',
    opacity=0.5,
    marker=dict(color='rgba(0, 0, 255, 0.5)') 
))

# Plot a red diagonal line representing perfect predictions (where actual = predicted).
min_val = min(y_test.min(), y_pred.min())
max_val = max(y_test.max(), y_pred.max())
fig.add_trace(go.Scatter(
    x=[min_val, max_val],
    y=[min_val, max_val],
    mode='lines',
    name='Perfect Prediction',
    line=dict(color='red', dash='dash')
))

# Adding axis labels and title for context, with matched style.
fig.update_layout(
    title='Actual vs Predicted Sales (Random Forest)',
    xaxis_title='Actual Sales',
    yaxis_title='Predicted Sales',
    legend_title='',
    title_x=0.5,
    template='simple_white',
    width=800,
    height=500,
    legend=dict(
        x=0.02,
        y=0.98,
        bgcolor='rgba(255,255,255,0.6)',
        bordercolor='LightGray',
        borderwidth=1
    )
)

# Display the plot.
fig.show()

# Save the plot for the dash app.
fig.write_html("assets/rf_actual_vs_pred.html")

For our second model, we used a Random Forest to better capture non-linear relationships and interactions between features. We maintained the same 80/20 train-test split as the linear regression model to ensure a fair comparison. The Random Forest performed significantly better, achieving an R² score of 0.77, meaning it explained about 77% of the variance in sales. The RMSE was approximately 218, indicating the model’s predictions were off by around 218 sales units on average. That’s a big improvement in both accuracy and variance explained compared to the linear regression baseline. These results suggest that Random Forest is a much stronger fit for this dataset, especially when dealing with complex patterns in the data.

In [54]:
# Similar to how we did with the linear regression model, we wanted to see a residual plot.
# This is an extra plot, we made it for our understanding, we will not be using it to present anything.
residuals = y_test - y_pred

# Create the residual plot.
fig = go.Figure()

# Add scatter points: predicted vs residuals.
fig.add_trace(go.Scatter(
    x=y_pred,
    y=residuals,
    mode='markers',
    name='Random Forest Residuals',
    opacity=0.5,
    marker=dict(color='rgba(0, 0, 255, 0.5)')
))

# Add horizontal reference line at y = 0.
fig.add_trace(go.Scatter(
    x=[y_pred.min(), y_pred.max()],
    y=[0, 0],
    mode='lines',
    name='Zero Error',
    line=dict(color='red', dash='dash')
))

# Match the style of previous plots exactly.
fig.update_layout(
    title='Residual Plot (Random Forest)',
    xaxis_title='Predicted Sales',
    yaxis_title='Residuals (Actual - Predicted)',
    title_x=0.5,
    template='simple_white',
    width=800,
    height=500,
    legend=dict(
        x=0.02,
        y=0.98,
        bgcolor='rgba(255,255,255,0.6)',
        bordercolor='LightGray',
        borderwidth=1
    )
)

# Show the plot.
fig.show()

# Plot shows better residual distribution than the other one (linear).

### **XGBoost**

In [55]:
# We want to take a closer look at XGBoost models.
# So here we want to start by finding the best parameters to tune the model.
# 1. Define the initial XGBoost model.
xgb = XGBRegressor(random_state=0, verbosity=0)

# 2. Set up a grid of hyperparameters to tune.
# We will test different tree depths, learning rates, and numbers of trees.
param_grid = {
    'n_estimators': [50, 75, 100, 150],
    'max_depth': [3, 5, 7, 10],
    'learning_rate': [0.01, 0.05, 0.1, 0.2]
}

# 3. Set up GridSearchCV to perform cross-validated search.
grid_search = GridSearchCV(
    estimator=xgb,
    param_grid=param_grid,
    scoring='neg_mean_squared_error',
    cv=10,
    verbose=1,
    n_jobs=-1
)

# 4. Fit the grid search on unscaled data.
# As with Random Forest, XGBoost is tree-based and does not require feature scaling.
# It handles numerical feature magnitude internally during tree construction.
grid_search.fit(X_train, y_train)

# 5. Print the best hyperparameters and best MSE (in positive form).
print("Best Parameters:", grid_search.best_params_)
print(f"Best MSE: {-grid_search.best_score_:.2f}")

Fitting 10 folds for each of 64 candidates, totalling 640 fits
Best Parameters: {'learning_rate': 0.1, 'max_depth': 5, 'n_estimators': 100}
Best MSE: 39984.10


In [56]:
# Refit the best XGBoost model using the full training set.
# This uses the best hyperparameters found by GridSearchCV.
best_xgb = grid_search.best_estimator_
best_xgb.fit(X_train, y_train)

# Make predictions on the test set using the refitted model.
y_pred = best_xgb.predict(X_test)

# Evaluate model performance on the test set.
# MSE: average squared error.
# RMSE: interpretable error in the same units as sales.
# R²: how much variance in sales the model explains.
mse = mean_squared_error(y_test, y_pred)
rmse = root_mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

# Print evaluation metrics.
print(f"Test MSE: {mse:.2f}")
print(f"Test RMSE: {rmse:.2f}")
print(f"Test R² Score: {r2:.3f}")

Test MSE: 49881.04
Test RMSE: 223.34
Test R² Score: 0.761


The XGBoost model achieved strong performance, with an R² score of 0.761, indicating it explains around 76 percent of the variance in sales. The RMSE of 223.34 shows that the model’s average prediction error is relatively low given the sales range. The MSE value of 49881.04 reinforces that overall error is well controlled, making this model a reliable option for predicting medication demand.

In [57]:
# Get feature importances from the best XGBoost model.
# These scores indicate how much each feature helped reduce prediction error across all trees.
# Features with higher scores consistently improved splits during training.

# Extract importance scores.
xgb_feature_importance = pd.Series(best_xgb.feature_importances_, index=X_train.columns)

# Select and sort the top 10 most important features.
top_features_xgb = xgb_feature_importance.nlargest(10).sort_values()

# Create horizontal bar chart using plotly.
fig = px.bar(
    x=top_features_xgb.values,
    y=top_features_xgb.index,
    orientation='h',
    labels={'x': 'Importance', 'y': 'Feature'},
    title='Top 10 Feature Importances (XGBoost)'
)

# Centre the title.
fig.update_layout(title_x=0.5)

# Show the plot.
fig.show()

# Save the plot for the dash app.
fig.write_html("assets/xgb_feature_importance.html")

The XGBoost feature importance plot highlights Season_Winter and Marketing_Spend as the top contributors to predicting sales, followed by Medication_Cough Suppressant and Season_Summer. This suggests that seasonal context and promotional effort (spend) are highly influential in the XGBoost model.

Comparing this to the Random Forest model, Marketing_Spend was by far the most dominant predictor, followed by Google_Trends and Medication_Cough Suppressant, with seasonal indicators like Season_Winter appearing further down.

While both models agree on the general importance of marketing spend and certain medications, XGBoost places greater weight on seasonality, especially winter, suggesting it may be better at capturing interaction effects or nonlinear seasonal patterns in the data.

In [58]:
# Create a scatter plot to compare actual vs. predicted sales.
# This helps us visually assess how well the model predictions align with the true values.
fig = go.Figure()

# Add scatter points for predicted vs actual values.
fig.add_trace(go.Scatter(
    x=y_test,
    y=y_pred,  
    mode='markers',
    name='XGBoost Predictions',
    opacity=0.5,
    marker=dict(color='rgba(0, 0, 255, 0.5)')  
))

# Plot a red diagonal line representing perfect predictions (where actual = predicted).
min_val = min(y_test.min(), y_pred.min())
max_val = max(y_test.max(), y_pred.max())
fig.add_trace(go.Scatter(
    x=[min_val, max_val],
    y=[min_val, max_val],
    mode='lines',
    name='Perfect Prediction',
    line=dict(color='red', dash='dash')
))

# Add axis labels and title for context, with matched style.
fig.update_layout(
    title='Actual vs Predicted Sales (XGBoost)',
    xaxis_title='Actual Sales',
    yaxis_title='Predicted Sales',
    legend_title='',
    title_x=0.5,
    template='simple_white',
    width=800,
    height=500,
    legend=dict(
        x=0.02,
        y=0.98,
        bgcolor='rgba(255,255,255,0.6)',
        bordercolor='LightGray',
        borderwidth=1
    )
)

# Display the plot.
fig.show()

# Save the plot for dash app.
fig.write_html("assets/xgb_actual_vs_pred.html")

This scatter plot shows how well the XGBoost model's predictions align with the actual sales values. Most points fall near the line, suggesting good performance overall. There is some spread at higher sales values, but the model still captures the overall trend effectively. This indicates that XGBoost is making reliable predictions across a wide range of sales values.

In [188]:
# Similar to how we did with the linear regression model, we wanted to see a residual plot.
# This is an extra plot, we made it for our understanding, we will not be using it to present anything.
residuals_xgb = y_test - y_pred  # these should be the XGBoost predictions

# Create the residual plot.
fig = go.Figure()

# Add scatter points: predicted vs residuals.
fig.add_trace(go.Scatter(
    x=y_pred,
    y=residuals_xgb,
    mode='markers',
    name='XGBoost Residuals',
    opacity=0.5,
    marker=dict(color='rgba(0, 0, 255, 0.5)')
))

# Add horizontal reference line at y = 0.
fig.add_trace(go.Scatter(
    x=[y_pred.min(), y_pred.max()],
    y=[0, 0],
    mode='lines',
    name='Zero Error',
    line=dict(color='red', dash='dash')
))

# Match the style of previous plots exactly.
fig.update_layout(
    title='Residual Plot (XGBoost)',
    xaxis_title='Predicted Sales',
    yaxis_title='Residuals (Actual - Predicted)',
    title_x=0.5,
    template='simple_white',
    width=800,
    height=500,
    legend=dict(
        x=0.02,
        y=0.98,
        bgcolor='rgba(255,255,255,0.6)',
        bordercolor='LightGray',
        borderwidth=1
    )
)

# Show the plot.
fig.show()

### **Feed Forward Neural Network**

In [15]:
# Time to use a feed forward neural network to predict medication demand. 
# Prepare features (everything except 'Date', 'Month' and 'Sales') and the target, 'Sales'.
X = df_clean.drop(columns=['Date', 'Month', 'Sales'])
y = df_clean['Sales']
# Categorical features must be one hot encoded.
X = pd.get_dummies(X, drop_first=True)
# Split data into training, test and validation data, with shuffling.
# First split will create the training data set (80%) and the test set (20%).
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, shuffle=True, random_state=0)
# Scale training and test features.
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [190]:
# WARNING: This takes a long time to run.
# Tuning code block. Re-run after changing parameters (e.g. learning rate, units, hidden layers, dropout, etc.) to find optimal epochs across folds.
# Define KFold object and empty lists.
kf = KFold(n_splits=10, shuffle=True, random_state=0)
val_mse_per_fold = []
epoch_per_fold = []

# Iterate through each fold (different validation fold each time) to retrieve optimal epoch number (at lowest MSE).
for train_idx, val_idx in kf.split(X_train_scaled):
    # Assign training and validation indices from overall training set.
    X_tr, X_val = X_train_scaled[train_idx], X_train_scaled[val_idx]
    y_tr, y_val = y_train.values[train_idx], y_train.values[val_idx]
    # Define model (now tuned). Two hidden layers with 128 and 64 units, with first hidden layer dropout of 0.2. 
    # One output unit due to regression task, using ReLU (linear identity) activation functions.
    model = keras.Sequential([
        keras.layers.Input(shape=(X_tr.shape[1],)),
        keras.layers.Dense(128, activation='relu'),
        keras.layers.Dropout(0.2),
        keras.layers.Dense(64, activation='relu'),
        keras.layers.Dense(1)
    ])
    # Complile model (now tuned). MSE used due to numerical outcome.
    model.compile(optimizer='adam', loss='mse', metrics=['mse'])
    # Train model (now tuned, except for epochs). Set maximum epochs at 450 for tuning. Batch size of 128.
    history = model.fit(
        X_tr, y_tr,
        validation_data=(X_val, y_val),
        epochs=450,
        batch_size=128,
        verbose=0
    )
    # Store best validation MSE and epoch at which it occurred.
    min_mse = min(history.history['val_mse'])
    best_epoch = np.argmin(history.history['val_mse']) + 1 
    # Append above values to list.
    val_mse_per_fold.append(min_mse)
    epoch_per_fold.append(best_epoch)

# Report averages of optimal validation MSE and epochs found across folds. 
print(f"Average minimum validation MSE across folds: {np.mean(val_mse_per_fold):.2f}")
print(f"Average best epoch: {np.mean(epoch_per_fold):.0f}")
# Average best epoch was 377, providing average MSE of 38373.72. Proceed with 385 epochs.

Average minimum validation MSE across folds: 38440.29
Average best epoch: 348


In [25]:
# Define final model architecture, same as tuned model, but now training on full training set (no validation).
nn = keras.Sequential([
    Input(shape=(X_train_scaled.shape[1],)),
    keras.layers.Dense(128, activation='relu'),
    keras.layers.Dropout(0.2),
    keras.layers.Dense(64, activation='relu'),
    keras.layers.Dense(1)
])

# Compile model.
nn.compile(optimizer='adam', loss='mse', metrics=['mse', RootMeanSquaredError(name='rmse'), 'mae'])

# Train model and capture training history. Note 385 epochs used.
history_final = nn.fit(
    X_train_scaled, y_train,
    epochs=385,
    batch_size=128,
    verbose=1
)

# Plot training MSE for reference (no validation set).
# Create the figure.
fig = go.Figure()
# Add training MSE trace.
fig.add_trace(go.Scatter(
    y=history_final.history['loss'],
    mode='lines',
    name='Train MSE'
))

# Customize layout.
fig.update_layout(
    title='Training MSE',
    xaxis_title='Epoch',
    yaxis_title='MSE',
    showlegend=True
)
fig.show()

# Evaluate tuned model on test set.
test_loss, test_mse, test_rmse, test_mae = nn.evaluate(X_test_scaled, y_test)
# Confirm order of above function assignment.
print(nn.metrics_names)
# Calculate average sales (target) in test set.
avg_test_sales = y_test.mean()
rmse_ratio = test_rmse / avg_test_sales
# Final metrics summary.
print(f"Test MSE: {test_loss:.2f}")
print(f"Test RMSE: {test_rmse:.2f}")
print(f"Test MAE: {test_mae:.2f}")
print(f"Test RMSE as % of Average Sales: {rmse_ratio:.2%}")

Epoch 1/385
46/46 ━━━━━━━━━━━━━━━━━━━━ 4s 10ms/step - loss: 711347.4375 - mae: 722.5178 - mse: 711347.4375 - rmse: 843.2329
Epoch 2/385
46/46 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - loss: 691547.2500 - mae: 712.0437 - mse: 691547.2500 - rmse: 831.4614
Epoch 3/385
46/46 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - loss: 626955.8125 - mae: 666.8622 - mse: 626955.8125 - rmse: 791.6744
Epoch 4/385
46/46 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - loss: 438590.1562 - mae: 544.4086 - mse: 438590.1562 - rmse: 661.9937
Epoch 5/385
46/46 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - loss: 256074.0469 - mae: 370.4747 - mse: 256074.0469 - rmse: 504.8385
Epoch 6/385
46/46 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - loss: 132814.5156 - mae: 229.4140 - mse: 132814.5156 - rmse: 363.7880
Epoch 7/385
46/46 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - loss: 99431.8047 - mae: 189.6612 - mse: 99431.8047 - rmse: 314.7536  
Epoch 8/385
46/46 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - loss: 87297.5312 - mae: 178.7421 - mse: 87297.5312 - rmse: 294.5202
Epoch 9/385
46/46

46/46 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - loss: 38659.8945 - mae: 96.4670 - mse: 38659.8945 - rmse: 194.6044
['loss', 'compile_metrics']
Test MSE: 39374.36
Test RMSE: 198.43
Test MAE: 92.11
Test RMSE as % of Average Sales: 26.77%


In [40]:
# Save trained model, scaler and input features for use in dash app.
nn.save("final_ffnn_model.keras")
import joblib
joblib.dump(scaler, "scaler.joblib")
input_columns = X.columns.tolist()
joblib.dump(input_columns, "input_columns.joblib")


['input_columns.joblib']

In [39]:
# Generate predictions using trained neural network. flatten() function is to convert from shape (n, 1) to (n,).
y_pred = nn.predict(X_test_scaled).flatten()  

# Plot actual vs predicted sales.
fig = go.Figure()
fig.add_trace(go.Scatter(
    x=y_test,
    y=y_pred,
    mode='markers',
    name='Neural Network Predictions',
    opacity=0.5
))
# Add ideal line: y = x.
min_val = np.min([y_test.min(), y_pred.min()])
max_val = np.max([y_test.max(), y_pred.max()])
fig.add_trace(go.Scatter(
    x=[min_val, max_val],
    y=[min_val, max_val],
    mode='lines',
    name='Ideal: y = x',
    line=dict(dash='dash', color='red')
))
# Layout settings.
fig.update_layout(
    title='Actual vs Predicted Sales (Neural Network)',
    xaxis_title='Actual Sales',
    yaxis_title='Predicted Sales',
    legend=dict(x=0.01, y=0.99),
    width=800,
    height=500,
    margin=dict(l=40, r=40, t=60, b=40)
)

# Display plot.
fig.show()

# Save plot for dash app.
fig.write_html("assets/nn_actual_vs_pred.html")

46/46 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step


In [27]:
# Additionally, calculate R^2 value for FFNN.
r2 = r2_score(y_test, y_pred)
print(f"R Squared Value: {r2:.2f}")

R Squared Value: 0.81


Test MSE: 39374.36 -
This is the average of squared differences between predicted and actual sales. Because it's in squared units, it's not easy to interpret directly — but it's useful for optimization and comparison.

Test RMSE: 198.43 -
This is the Root Mean Squared Error, in the same units as 'Sales'. This tells us that on average, the model’s predictions are off by about 198 units of sales. This is easier to understand than MSE, a good metric for business or forecasting stakeholders.

Test MAE: 92.11 - 
This is the Mean Absolute Error, or average of absolute differences. It tells us that regardless of whether it's over or under, the model is off by ~92 units of sales on average. Note that unlike RMSE, MAE doesn't punish large errors as harshly — it gives a more balanced view of general performance.

RMSE as % of Average Sales: 26.77% -
This tells us how big the typical error is relative to average sales, meaning that the model’s typical prediction error is about 27% of the average sales value.

Therefore, the FFNN displayed the best performance thus far.

### **SARIMA Time Series Models**
Time series forecasting is a technique that uses historical data points, ordered over time, to predict future values. In this analysis, we applied time series modeling to predict sales of over-the-counter medications across regions and product types.

We used the SARIMA (Seasonal AutoRegressive Integrated Moving Average) model, which accounts for:
- **Trends** over time (e.g., increasing or decreasing sales),
- **Seasonality** (e.g., weekly or monthly patterns),
- **Autocorrelation** (how past values influence future values).

#### Time Series 1: Future Forecast (Next 30 Days)

In [62]:
# Convert 'Date' to datetime and aggregate sales across all medications and regions by date.
df['Date'] = pd.to_datetime(df['Date'])
df_filtered = df.copy()
df_grouped = df_filtered.groupby('Date')['Sales'].sum().asfreq('D').ffill()

# Use entire time series to train the model.
train = df_grouped

# Fit SARIMA model with weekly seasonality.
model = SARIMAX(train, order=(1, 1, 1), seasonal_order=(1, 1, 1, 7))
results = model.fit(disp=False)

# Forecast the next 30 days. 
date_forecast = results.get_forecast(steps=30)
forecast_index = pd.date_range(start=train.index[-1] + timedelta(days=1), periods=30)
forecast_series = pd.Series(date_forecast.predicted_mean.values, index=forecast_index)
conf_int = date_forecast.conf_int()

# Plot with plotly.
fig = go.Figure()
fig.add_trace(go.Scatter(x=train.index, y=train.values, mode='lines', name='Historical'))
fig.add_trace(go.Scatter(x=forecast_index, y=forecast_series, mode='lines', name='Forecast'))
fig.add_trace(go.Scatter(
    x=forecast_index.tolist() + forecast_index[::-1].tolist(),
    y=conf_int.iloc[:, 0].tolist() + conf_int.iloc[:, 1][::-1].tolist(),
    fill='toself',
    fillcolor='rgba(255,165,0,0.3)',
    line=dict(color='rgba(255,255,255,0)'),
    hoverinfo="skip",
    name='Confidence Interval'
))
fig.update_layout(
    title="Overall Forecast for Next 30 Days",
    xaxis_title="Date",
    yaxis_title="Sales",
    template="plotly_white"
)
fig.show()


The model was used to forecast total sales 30 days into the future using the full historical dataset.
As the future values do not exist in our dataset, we evaluated the forecast based on how well it aligns with past trends and the width of the confidence interval.
The results show a stable continuation of existing seasonal patterns, with forecast uncertainty gradually increasing — which is expected in time series modeling.

#### Time Series 2: Forecast – Overall with Evaluation
This model forecasts total daily sales across all medications and regions using a SARIMA model with weekly seasonality.
To assess accuracy, the last 30 days of data were held out as a test set.
Forecast performance was evaluated using MSE, RMSE, and MAE, showing how well the model predicts actual sales it was not trained on.

Evaluation Metrics:

MSE: 2709807.14

RMSE: 5204.79

MAE: 5057.70

Relative RMSE: ~2.63% of average daily sales.

The model performed strongly, with an RMSE of just 5,200 units per day — only 2.63% of the average daily sales volume (~14,674).
This indicates that the model captured overall demand trends and weekly seasonality effectively.
While some variation is expected during seasonal peaks or unexpected demand shifts, the model provides a relatively reliable short-term forecast.

In [64]:
# Reuse the grouped data.
train = df_grouped[:-30]
test = df_grouped[-30:].dropna() 

# Forecast 30 days.
model = SARIMAX(train, order=(1, 1, 1), seasonal_order=(1, 1, 1, 7))
results = model.fit(disp=False)
forecast = results.get_forecast(steps=30)
forecast_index = pd.date_range(start=train.index[-1] + timedelta(days=1), periods=30)
forecast_series = pd.Series(forecast.predicted_mean.values, index=forecast_index)
conf_int = forecast.conf_int()

# Align test to forecast index (may shrink test if any mismatch).
test = test.reindex(forecast_index)

# Metrics.
mse = mean_squared_error(test, forecast_series)
rmse = np.sqrt(mse)
mae = mean_absolute_error(test, forecast_series)
print(f"Overall Forecast → MSE: {mse:.2f}, RMSE: {rmse:.2f}, MAE: {mae:.2f}")

# Plot.
fig = go.Figure()
fig.add_trace(go.Scatter(x=train.index, y=train.values, mode='lines', name='Train'))
fig.add_trace(go.Scatter(x=test.index, y=test.values, mode='lines', name='Actual'))
fig.add_trace(go.Scatter(x=forecast_index, y=forecast_series, mode='lines', name='Forecast'))
fig.add_trace(go.Scatter(
    x=forecast_index.tolist() + forecast_index[::-1].tolist(),
    y=conf_int.iloc[:, 0].tolist() + conf_int.iloc[:, 1][::-1].tolist(),
    fill='toself',
    fillcolor='rgba(255,165,0,0.3)',
    line=dict(color='rgba(255,255,255,0)'),
    hoverinfo="skip",
    name='Confidence Interval'
))
fig.update_layout(
    title="Overall Forecast with Historical Test Set",
    xaxis_title="Date",
    yaxis_title="Sales",
    template="plotly_white"
)
fig.show()

# Save the plot for dash app.
fig.write_html("assets/overall_forecast_with_testset.html")

Overall Forecast → MSE: 27089801.05, RMSE: 5204.79, MAE: 5057.70


### Time Series 3: Forecast – Per Medication and Region
This section trains separate SARIMA models for each medication-region pair.
It calculates forecast accuracy (MSE, RMSE, MAE) for each combination.

In [65]:
# Grouped by medication and region.
medications = df['Medication'].unique()
regions = df['Region'].unique()
forecast_periods = 30
metrics = []

# Loop through each medication and region combination.
for med in medications:
    for region in regions:
        try:
            # Filter and clean.
            df_filtered = df[(df['Medication'] == med) & (df['Region'] == region)].copy()
            df_filtered['Date'] = pd.to_datetime(df_filtered['Date'])
            df_grouped = df_filtered.groupby('Date')['Sales'].sum().asfreq('D').ffill()

            if df_grouped.isna().sum() > 0 or len(df_grouped) < 2 * forecast_periods:
                print(f"Skipping {med} in {region} — insufficient or missing data.")
                continue

            # Train-test split.
            train = df_grouped[:-forecast_periods]
            test = df_grouped[-forecast_periods:]

            # Fit SARIMA.
            model = SARIMAX(train, order=(1, 1, 1), seasonal_order=(1, 1, 1, 7))
            results = model.fit(disp=False)
            forecast = results.get_forecast(steps=forecast_periods)
            forecast_index = pd.date_range(start=train.index[-1] + timedelta(days=1), periods=forecast_periods)
            forecast_series = pd.Series(forecast.predicted_mean.values, index=forecast_index)
            conf_int = forecast.conf_int()

            # Align test and forecast for metrics.
            test = test.reindex(forecast_index)
            if test.isna().sum() > 0:
                print(f"Skipping {med} in {region} — NaNs in test after reindexing.")
                continue

            # Metrics.
            mse = mean_squared_error(test, forecast_series)
            rmse = np.sqrt(mse)
            mae = mean_absolute_error(test, forecast_series)
            print(f"{med} | {region} → MSE: {mse:.2f}, RMSE: {rmse:.2f}, MAE: {mae:.2f}")
            metrics.append({"Medication": med, "Region": region, "MSE": mse, "RMSE": rmse, "MAE": mae})

            # Plot.
            fig = go.Figure()
            fig.add_trace(go.Scatter(x=train.index, y=train.values, mode='lines', name='Train'))
            fig.add_trace(go.Scatter(x=test.index, y=test.values, mode='lines', name='Actual'))
            fig.add_trace(go.Scatter(x=forecast_index, y=forecast_series, mode='lines', name='Forecast'))
            fig.add_trace(go.Scatter(
                x=forecast_index.tolist() + forecast_index[::-1].tolist(),
                y=conf_int.iloc[:, 0].tolist() + conf_int.iloc[:, 1][::-1].tolist(),
                fill='toself',
                fillcolor='rgba(255,165,0,0.3)',
                line=dict(color='rgba(255,255,255,0)'),
                hoverinfo="skip",
                name='Confidence Interval'
            ))
            fig.update_layout(
                title=f"{med} Sales Forecast - {region}",
                xaxis_title="Date",
                yaxis_title="Sales",
                template="plotly_white"
            )

            fig.show()

        except Exception as e:
            print(f"Failed to forecast {med} in {region}: {e}")

# Results.
forecast_metrics = pd.DataFrame(metrics)
print(forecast_metrics)



Cold Relief | Toronto → MSE: 720278.76, RMSE: 848.69, MAE: 810.88


Cold Relief | Vancouver → MSE: 300561.23, RMSE: 548.23, MAE: 485.38


Cold Relief | Montreal → MSE: 772329.28, RMSE: 878.82, MAE: 851.43


Cold Relief | Calgary → MSE: 637424.11, RMSE: 798.39, MAE: 772.96


Allergy Relief | Toronto → MSE: 5669.11, RMSE: 75.29, MAE: 61.43


Allergy Relief | Vancouver → MSE: 11980.78, RMSE: 109.46, MAE: 72.89


Allergy Relief | Montreal → MSE: 23879.88, RMSE: 154.53, MAE: 81.03


c:\Users\tenzi\anaconda3\Lib\site-packages\statsmodels\base\model.py:607: ConvergenceWarning:

Maximum Likelihood optimization failed to converge. Check mle_retvals



Allergy Relief | Calgary → MSE: 6806.14, RMSE: 82.50, MAE: 66.60


Pain Relief | Toronto → MSE: 72057.62, RMSE: 268.44, MAE: 128.79


Pain Relief | Vancouver → MSE: 17912.69, RMSE: 133.84, MAE: 96.65


Pain Relief | Montreal → MSE: 16201.78, RMSE: 127.29, MAE: 91.80


Pain Relief | Calgary → MSE: 11438.51, RMSE: 106.95, MAE: 70.97


Cough Suppressant | Toronto → MSE: 1022687.13, RMSE: 1011.28, MAE: 737.26


Cough Suppressant | Vancouver → MSE: 260530.00, RMSE: 510.42, MAE: 424.57


Cough Suppressant | Montreal → MSE: 304140.45, RMSE: 551.49, MAE: 529.66


Cough Suppressant | Calgary → MSE: 104786.42, RMSE: 323.71, MAE: 242.21


Fever Reducer | Toronto → MSE: 276307.61, RMSE: 525.65, MAE: 437.92


Fever Reducer | Vancouver → MSE: 51499.31, RMSE: 226.93, MAE: 203.83


c:\Users\tenzi\anaconda3\Lib\site-packages\statsmodels\base\model.py:607: ConvergenceWarning:

Maximum Likelihood optimization failed to converge. Check mle_retvals



Fever Reducer | Montreal → MSE: 182791.48, RMSE: 427.54, MAE: 406.95


Fever Reducer | Calgary → MSE: 149384.53, RMSE: 386.50, MAE: 360.49


           Medication     Region           MSE         RMSE         MAE
0         Cold Relief    Toronto  7.202788e+05   848.692380  810.880937
1         Cold Relief  Vancouver  3.005612e+05   548.234648  485.382911
2         Cold Relief   Montreal  7.723293e+05   878.822669  851.433990
3         Cold Relief    Calgary  6.374241e+05   798.388448  772.955619
4      Allergy Relief    Toronto  5.669112e+03    75.293506   61.426104
5      Allergy Relief  Vancouver  1.198078e+04   109.456731   72.889332
6      Allergy Relief   Montreal  2.387988e+04   154.531172   81.030712
7      Allergy Relief    Calgary  6.806140e+03    82.499332   66.597500
8         Pain Relief    Toronto  7.205762e+04   268.435506  128.794307
9         Pain Relief  Vancouver  1.791269e+04   133.838294   96.653168
10        Pain Relief   Montreal  1.620178e+04   127.286227   91.797236
11        Pain Relief    Calgary  1.143851e+04   106.950971   70.965635
12  Cough Suppressant    Toronto  1.022687e+06  1011.279948  737

In [66]:
# Interpretations based on medication type.
forecast_metrics["Interpretation"] = forecast_metrics.apply(lambda row: {
    # Cold Relief.
    ("Cold Relief", "Toronto"): "Forecasting was challenging due to sharp winter demand spikes (RMSE: 849).",
    ("Cold Relief", "Vancouver"): "Moderate accuracy; fewer extreme seasonal peaks contributed to lower error (RMSE: 548).",
    ("Cold Relief", "Montreal"): "Frequent cold-season surges led to high forecast error (RMSE: 879).",
    ("Cold Relief", "Calgary"): "Forecasts were moderately accurate; demand fluctuated during winter months (RMSE: 798).",

    # Allergy Relief.
    ("Allergy Relief", "Toronto"): "Forecasts were highly accurate, supported by consistent spring allergy patterns (RMSE: 75).",
    ("Allergy Relief", "Vancouver"): "Spring-related demand was well captured, resulting in low error (RMSE: 109).",
    ("Allergy Relief", "Montreal"): "Slightly higher variability in allergy season led to moderate forecast error (RMSE: 155).",
    ("Allergy Relief", "Calgary"): "Clear seasonal trends supported strong model performance (RMSE: 82).",

    # Pain Relief.
    ("Pain Relief", "Toronto"): "Mild fluctuations in daily use contributed to modest error (RMSE: 268).",
    ("Pain Relief", "Vancouver"): "Stable demand throughout the year allowed for reliable forecasts (RMSE: 134).",
    ("Pain Relief", "Montreal"): "Minimal seasonality supported accurate model results (RMSE: 127).",
    ("Pain Relief", "Calgary"): "Very strong performance driven by steady usage patterns (RMSE: 107).",

    # Cough Suppressant.
    ("Cough Suppressant", "Toronto"): "Unpredictable seasonal spikes resulted in the highest forecast error (RMSE: 1011).",
    ("Cough Suppressant", "Vancouver"): "Moderate accuracy; seasonal variation contributed to forecast noise (RMSE: 510).",
    ("Cough Suppressant", "Montreal"): "Cold-season demand was variable, reducing forecast reliability (RMSE: 551).",
    ("Cough Suppressant", "Calgary"): "Lower error compared to other regions; seasonal shifts were more predictable (RMSE: 324).",

    # Fever Reducer.
    ("Fever Reducer", "Toronto"): "Forecasts were affected by flu season peaks, particularly in winter (RMSE: 526).",
    ("Fever Reducer", "Vancouver"): "Low forecast error due to relatively stable demand (RMSE: 226).",
    ("Fever Reducer", "Montreal"): "Moderate accuracy; some demand variability during peak illness periods (RMSE: 427).",
    ("Fever Reducer", "Calgary"): "Consistent performance; mild seasonal influence captured effectively (RMSE: 387)."

}.get((row["Medication"], row["Region"]), "Interpretation not available."), axis=1)


# Table.
from IPython.display import display
display(forecast_metrics.style.set_caption("Forecast Accuracy by Medication and Region").format(precision=2))
